### ISSUE TITLE
Lazy load url for download

### ISSUE THREAD

**Issue by kgrevehagen** [google/ExoPlayer#8324]
In order for me to get a url for downloading content, I need to send a request with an id to our servers. When queuing many tracks for download, this does not work quite well(urls would expire, it would take too much time up front etc), so I need a system that lazily sends the said request when it is time to actually download each track.

I can't find an out-of-the-box solution to accomplish this. I had the same issue with playback, so I created a custom `MediaSource`(using `Loader`/`Loadable`) that does the trick. So I came up with a similar solution that seems to work, but I want to know if it makes sense or if there are better ways to do it.

My first solution was to manually create a `DownloadHelper` with a `MediaSource`(reusing my custom `MediaSource` implementation). This is not perfect, because my `MediaSource` would prepare outside of the `DownloadService`(by calling `DownloadHelper.prepare()`), which obviously is not the best.

My so far best solution:
Create a custom `DownloaderFactory` and in turn a custom `Downloader` that fetches the url before it passes that url to the correct downloader(e.g. `DashDownloader` or `ProgressiveDownloader`). By doing this I need to work with the `DownloadManager` manually by adding empty `MediaItem`s(except they have my id so I can get it in the `Downloader`) and I would have to start the `DownloadService` manually, which is not a big problem, imo.

1. I wonder how I should actually send the request. `SegmentDownloader` uses `Executor`, `RunnableFutureTask`, `ParsingLoadable` and also handles `PriorityTaskManager` etc. Do I need all of this, or could I just send the request directly inside `download()` of my custom `Downloader`?
2. Is it fine to manually handle the `DownloadManager` and starting the `DownloadService`(not by using `sendAddDownload()`, but rather `sendResumeDownloads()` after calling `downloadManager.addDownload`)?
3. This means I do not follow your documentation on exoplayer.dev, e.g. by creating a `DownloadManager` with my own `DownloaderFactory` and not using any of the stuff in the `DownloaderHelper`.

Do the latter sound reasonable, or are there any other solutions that would make more sense?
Hope this was not too confusing, thanks in advance!

**Comment by marcbaechinger** [comment]
>  are there any other solutions that would make more sense?

Sorry when not answering directly your other questions directly for now, just checking :

When you [create your subclass of a `DownloadService`](https://exoplayer.dev/downloading-media.html#creating-a-downloadservice) service that you need, you need to implement `getDownloadManager()`. If you construct your download manager you need to pass a `DataSource.Factory`. You can use a [`ResolvingDataSource.Factory`](https://github.com/google/ExoPlayer/blob/release-v2/library/core/src/main/java/com/google/android/exoplayer2/upstream/ResolvingDataSource.java) which has a `Resolver` that you implement. That's how you can intercept the network call for downloading and replace the uri with the new one that you get from your server. 

This method is called when the download for a given media starts, and you can also block the resolving call so you can ask your server just at the moment before downloading:

```
/**
     * Resolves a {@link DataSpec} before forwarding it to the wrapped {@link DataSource}. This
     * method is allowed to block until the {@link DataSpec} has been resolved.
     *
     * <p>Note that this method is called for every new connection, so caching of results is
     * recommended, especially if network operations are involved.
     *
     * @param dataSpec The original {@link DataSpec}.
     * @return The resolved {@link DataSpec}.
     * @throws IOException If an {@link IOException} occurred while resolving the {@link DataSpec}.
     */
    DataSpec resolveDataSpec(DataSpec dataSpec) throws IOException;
```

Would that work? I probably haven't understood all the details of you requirements, but itseems to me that should work for 'In order for me to get a url for downloading content, I need to send a request with an id to our servers.'

**Comment by kgrevehagen** [comment]
In a "normal" scenario that would work, but I see now that I wasn't clear enough in my first question. When I ask my server for the url, I get the url back with information if it is progressive, dash, hls or whatever. If I use the `ResolvingDataSource`, I would have already had to decide this, but I can't until I get the response back from the server. This is why I couldn't use `ResolvingDataSource` for playback, I had to create a custom `MediaSource`. And I think this is the same reason I can't use `ResolvingDataSource` here as well, it would be a little too late in the process. Therefore, I suspect I need to create a custom `Downloader`, just like I did with a custom `MediaSource`.

Something like this:
```
val downloader
val (url, type) = fetchInfoFromServer()
if (type == progressive){
  downloader = ProgressiveDownloader(...)
} else if (type == dash) {
  downloader = DashDownloader(...)
}
downloader.download(...)
```


If my assumption of creating a custom `Downloader` is correct, my question would then be how to fetch the information from my server. Can I just do it directly/blocking or do I have to implement anything like you have done in the other downloaders(`Executor`, `RunnableFutureTask` etc...)? What would the pros/cons be of the two?
